api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,387 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return referencePlugin.generateOSBResponse(ctx, UpdateService, nil)
+}
+
+func updateSharedInstance(ctx context.Context, repository storage.Repository, req *web.Request, instance *types.ServiceInstance, next web.Handler) (*web.Response, error) {
+	err := isValidSharedInstancePatchRequest(ctx, repository, req, instance)
+	if err != nil {
+		return nil, err
+	}
+	return next.Handle(req)
+}
+
+// Bind intercepts bind requests and check if the instance is in the platform from where the request comes
+func (referencePlugin *referenceInstancePlugin) Bind(req *web.Request, next web.Handler) (*web.Response, error) {
+	return referencePlugin.handleBinding(req, next)
+}
+
+// Unbind intercepts unbind requests and check if the instance is in the platform from where the request comes
+func (referencePlugin *referenceInstancePlugin) Unbind(req *web.Request, next web.Handler) (*web.Response, error) {
+	return referencePlugin.handleBinding(req, next)
+}
+
+// FetchBinding intercepts get service binding requests and check if the instance owner is the same as the one requesting the operation
+func (referencePlugin *referenceInstancePlugin) FetchBinding(req *web.Request, next web.Handler) (*web.Response, error) {
+	return referencePlugin.handleBinding(req, next)
+}
+
+func (referencePlugin *referenceInstancePlugin) FetchService(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	instanceID := req.PathParams["instance_id"]
+	dbInstanceObject, err := storage.GetObjectByField(ctx, referencePlugin.repository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := storage.IsReferencePlan(ctx, referencePlugin.repository, types.ServicePlanType.String(), "id", instance.ServicePlanID)
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return referencePlugin.generateOSBResponse(ctx, FetchService, instance)
+}
+
+func (referencePlugin *referenceInstancePlugin) generateOSBResponse(ctx context.Context, method OperationCategory, instance *types.ServiceInstance) (*web.Response, error) {
+	var marshal []byte
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	switch method {
+	case Provision:
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusCreated,
+			Header:     headers,
+		}, nil
+	case FetchService:
+		osbResponse, err := referencePlugin.buildOSBFetchServiceResponse(ctx, instance)
+		if err != nil {
+			return nil, err
+		}
+		marshal, err = json.Marshal(osbResponse)
+		if err != nil {
+			return nil, err
+		}
+		return &web.Response{
+			Body:       marshal,
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	default:
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	}
```

Comment:
is this needed here? that's not common logic, it is specific to the different types of request. I think it would make more sense to have the specific body/status codes passed in from the different flows, and only have the common logic here (headers). It would result in less code and better readability.